feat: env var to control exception wrapping in tests

james-garner-canonical opened the pull request:
This PR adds a new environment variable, `SCENARIO_BARE_CHARM_ERRORS`. If this is set to `1` or `true` (case insensitive), then exceptions raised in charm code in state-transition tests are re-raised as-is. If unset or set to any other value, then the default behaviour matches the current behaviour: charm errors are wrapped in `UncaughtCharmError` and that is raised instead.

The idea would be to make `SCENARIO_BARE_CHARM_ERRORS=1` the default behaviour in a future major release. In the meantime, the environment variable allows charms to explicitly opt into the new behaviour now, for a better debugging experience, or explicitly lock in the current behaviour to avoid future breakages.

The name is up for debate, but was chosen to follow the format of `SCENARIO_SKIP_CONSISTENCY_CHECKS`.

The `testing` tests have been updated to exercise both behaviours, while the `ops` tests continue to assume the existing behaviour, and will require updates when we eventually change the default.

`tox.ini` has been updated to always pass the scenario environment variables along.

Resolves #1428.

benhoyt commented on testing/src/scenario/_environ.py:
```diff
@@ -0,0 +1,10 @@
+import os
+
+
+def wrap_charm_errors() -> bool:
```
I think it's quite unusual for a boolean environment variable to be true by default (in fact, often setting it means true regardless of value). So I'd prefer we invert the name and sense, perhaps `SCENARIO_BARE_CHARM_ERRORS`.

james-garner-canonical replied:
This is forward-looking naming, since the intention is that the default will eventually be false, but I don't think it's a big deal either way. Happy for it to start out named such that the default is false, and eventually switch to being true by default.

benhoyt commented on testing/src/scenario/_environ.py:
```diff
@@ -0,0 +1,10 @@
+def wrap_charm_errors() -> bool:
+    """Return whether scenario should wrap charm errors with ``UncaughtCharmError``."""
+    default = 'True'
+    return os.getenv('SCENARIO_WRAP_CHARM_ERRORS', default).capitalize() != 'False'
```
In my experience it's more typical to use `=1` rather than `=true` (or `=false`) for environment variables. Also, nit, but I don't think we need to name the *default* variable. Let's do this:

```python
return os.getenv('SCENARIO_BARE_CHARM_ERRORS', 'false') in ['1', 'true']
```

james-garner-canonical replied:
Good idea to support additional booly values like `0` and `1`.

I named the default to call attention to it, since that's the line that will change in future -- but I don't mind inlining it.

james-garner-canonical replied:
This is now in-lined at the point where we raise `UncaughtCharmError`.

james-garner-canonical commented on testing/tests/test_context.py:
```diff
@@ -9,6 +9,7 @@
 from scenario import Context, State
 from scenario.errors import UncaughtCharmError
 from scenario.state import _Event, _next_action_id
+from scenario._environ import wrap_charm_errors
```
> I don't think we should be pulling in this private code in tests.

I figured it was fine since scenario's tests already use its private, internal things (e.g. the line above).

> We could just have the test read the environment variable directly. This is too dependent on the exact structure of the implementation.

The intention was to make the tests *less* dependent on the environment variable name, default, and semantics.  I dislike having the tests ready the environment directly for that reason. But if we switch to them only writing to the environment as you suggest, then the usefulness of the abstraction goes away.

> However, I think we should do it an alternative way: keep the tests which test the existing behaviour, and add a test or two that sets `SCENARIO_BARE_CHARM_ERRORS=1` and tests that behaviour too. This would also simplify running the tests.

Good idea. That just means we'll be modifying all the tests that assume `UncaughtCharmError` later when we change the default value of the environment variable, but that's not too bad.

> In addition, I don't think we should have a whole separate _environ module just for this one-liner helper. I reckon just put it in _runtime.py.

If we're not using it in tests, then this makes sense, it's where I had the logic initially -- we don't even need the helper if the tests don't need to use the logic, so I'll just in-line it all.

james-garner-canonical commented on .github/workflows/framework-tests.yaml:
```diff
@@ -33,6 +33,7 @@ jobs:
       matrix:
         os: [ubuntu-latest, macos-latest]
         python-version: ["3.10", "3.12", "3.14"]
+        wrap-charm-errors: ["True", "False"]
```
I've reverted the changes to this file, so there are no CI changes now.

james-garner-canonical commented:
I've refactored as per @benhoyt's suggestions:
- Dropped the CI changes
- Inverted the name and default (now `SCENARIO_BARE_CHARM_ERRORS` defaulting to false)
- Dropped the helper function and in-lined the logic at the place we raise `UncaughtCharmError`

I've parametrized the existing tests that caught `UncaughtCharmError` to set the environment variable using `monkeypatch.setenv`, which both exercises the environment variable logic and ensures that our expectations about the shape and type of exceptions hold in both cases. I think this is the right approach for the internal `testing` tests, but I'm less sure about the noise it adds to `test/test_model_relation_data_class.py` -- maybe the `ops` tests should just assume the current default.

dimaqq reviewed (commented):
How will this env var be documented?

The other env var used to be documented: https://pypi.org/project/ops-scenario/4.0.4.1/
It's no longer documented today.

TheJJ commented:
I've implemented the alternative approach that avoids environment variables, and instead controls this through `Context` in #2145.

james-garner-canonical commented:
I lean towards the environment variable as I think it will give us a better pathway to changing the default behavior in future, but I'm definitely open to discussing alternatives.


Regarding the developer experience, it's worth noting that you can still define the behavior in your tests with an environment variable, for example with an autouse fixture. To control the behavior per-test with relatively low boilerplate you could instead use pytest.mark.usefixtures.

tonyandrewmeyer reviewed (approved):
Thanks! I think this will be quite nice for many people.

Everything seems good to me, but I think we should expand the tested cases. Approving optimistically.

james-garner-canonical commented on testing/tests/test_context.py:
```diff
@@ -96,7 +96,8 @@ def test_app_name_and_unit_id():
 
 
-def test_context_manager_uncaught_error():
+@pytest.mark.parametrize('bare_charm_errors', ('1', ''))
```
I'm OK with making this change, but I'll wait to see what @benhoyt thinks. I also wonder if we could just have one function that explicitly tests all those variants, and leave the rest as-is.

tonyandrewmeyer replied:
One test that explicitly tests all of those would be fine with me.

james-garner-canonical commented on testing/src/scenario/_runtime.py:
```diff
@@ -331,6 +331,8 @@ def exec(
             except (NoObserverError, ActionFailed):
                 raise  # propagate along
             except Exception as e:
+                if os.getenv('SCENARIO_BARE_CHARM_ERRORS', 'false').lower() in {'1', 'true'}:
```
Will leave it as is.

james-garner-canonical commented on testing/tests/test_e2e/test_relations.py:
```diff
@@ -623,7 +623,8 @@ def test_broken_relation_not_in_model_relations(mycharm):
 
 
-def test_get_relation_when_missing():
+@pytest.mark.parametrize('bare_charm_errors', ('1', ''))
```
An alternative option for env var proofing would be if these tests just became a bit looser, they could all use something like this instead of parametrizing:
```py
with pytest.raises((UncaughtCharmError, TheActualError)):
```
Though I guess you'd still need to get conditional to check any of the error contents ...

james-garner-canonical replied:
I'm now wondering if it would be even cleaner to just `monkeypatch.setenv('SCENARIO_BARE_CHARM_ERRORS', 'false')`  in the tests where we don't care about testing both behaviours to make explicit which behaviour we're testing. I'll iterate on this next week.

dimaqq commented on testing/tests/test_runtime.py:
```diff
@@ -118,7 +121,8 @@ def test_env_clean_on_charm_error():
     # Ensure that some other error didn't occur (like AssertionError!).
-    assert 'ZeroDivisionError' in str(exc.value)
+    exc = exc_info.value if bare_charm_errors else exc_info.value.__cause__
+    assert isinstance(exc, ZeroDivisionError)
```
+1 on the discussion in the other comment that the code flow should not depend on the value of the test parameter.

Then most tests could read something like:

```suggestion
    assert isinstance(exc, ZeroDivisionError) or isinstance(exc_info.value.__cause__, ZeroDivisionError)
```

dimaqq commented on testing/tests/test_runtime.py:
```diff
@@ -153,13 +158,50 @@ def _on_start(self, _: ops.StartEvent):
+        (ValueError, '1'),
+        (ValueError, 'tRuE'),  # case insensitive
+        (UncaughtCharmError, ''),
+        (UncaughtCharmError, '✨ aNy ✩ sTrInG ✧ ReAlLy ...'),
```
Nice, but maybe too clever.

How about:

```suggestion
        (ValueError, '1'),
        (ValueError, 'True'),
        (ValueError, 'true'),
        (UncaughtCharmError, '0'),
        (UncaughtCharmError, 'False'),
        (UncaughtCharmError, 'false'),
        (UncaughtCharmError, ''),
        (UncaughtCharmError, 'anything else'),
```

I would also swap them around (setting, observable_effect).

Aside: this doesn't test the missing env var case, though I imagine that's covered by regular tests.

james-garner-canonical commented:
@dimaqq do you want to review this PR further, or are you happy for it to be merged?

dimaqq commented on testing/tests/test_runtime.py:
```diff
@@ -163,3 +165,47 @@ def _on_start(self, _: ops.StartEvent):
+        # any other value is treated as logically false
+        (UncaughtCharmError, '0'),
+        (UncaughtCharmError, 'false'),
+        (UncaughtCharmError, '11'),
```
I would expect `"11"` to be truthy.

dimaqq replied:
That being said, I'm happy to go with a majority vote.

james-garner-canonical replied:
@benhoyt I went with your initial suggestion of testing for 1 and true, but do you think it's worth expanding this at all?

james-garner-canonical replied:
@benhoyt I went with your initial suggestion of testing for 1 and true, but do you think it's worth expanding this at all?

james-garner-canonical replied:
Per discussion with @benhoyt on Monday, I've updated the implementation to treat positive integers as true as well, and updated the test cases accordingly.
